source/fab/steps/link_exe.py

```diff
@@ -8,17 +8,31 @@
 
 """
 import logging
+import os
 from string import Template
 from typing import List
 
+from fab import artefacts
 from fab.constants import BUILD_OUTPUT
-from fab.steps import Step
+from fab.steps import Step, archive_objects
 from fab.util import log_or_dot, run_command
-from fab.artefacts import ArtefactsGetter, CollectionConcat
+from fab.artefacts import ArtefactsGetter, CollectionGetter
 
 logger = logging.getLogger(__name__)
 
-DEFAULT_SOURCE_GETTER = CollectionConcat(['compiled_c', 'compiled_fortran'])
+
+class DefaultLinkerSource(ArtefactsGetter):
+    """
+    A source getter specifically for linking.
+    Looks for the default output from ArchiveObjects, falls back to artefacts.CompiledFortranAndC.
+    This allows a link step to work with or without a preceding object archive step.
+    """
+    def __call__(self, artefact_store):
+        return CollectionGetter(archive_objects.DEFAULT_COLLECTION_NAME)(artefact_store) \
+               or artefacts.CompiledFortranAndC(artefact_store)
```

Comment:
How does this cope with multiple executables not using all the compiled objects?

Reply:
We've not covered that use case yet but a recent review did raise [a ticket for it](https://github.com/metomi/fab/projects/8#card-82482567).